File: data_structures/circular_buffer/matlab/test_circular_buffer.m
% test_circular_buffer.m — 循環バッファのテストスクリプト

b = circular_buffer('new', 3);
assert(circular_buffer(b, 'is_empty'));
assert(circular_buffer(b, 'capacity') == 3);

b = circular_buffer(b, 'push', 10);
b = circular_buffer(b, 'push', 20);
b = circular_buffer(b, 'push', 30);
assert(circular_buffer(b, 'is_full'));
assert(circular_buffer(b, 'peek') == 10);

[b, v] = circular_buffer(b, 'pop');
assert(v == 10);
[b, v] = circular_buffer(b, 'pop');
assert(v == 20);

b = circular_buffer(b, 'push', 40);
b = circular_buffer(b, 'push', 50);
assert(circular_buffer(b, 'is_full'));
assert(circular_buffer(b, 'peek') == 30);

try
    circular_buffer(b, 'push', 99);
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'circular_buffer:overflow'));
end

[b, ~] = circular_buffer(b, 'pop');
[b, ~] = circular_buffer(b, 'pop');
[b, ~] = circular_buffer(b, 'pop');
assert(circular_buffer(b, 'is_empty'));

try
    circular_buffer(b, 'pop');
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'circular_buffer:underflow'));
end

disp('All tests passed.');

File: data_structures/circular_buffer/python/test_circular_buffer.py
import pytest
from circular_buffer import CircularBuffer


def test_new_is_empty():
    assert CircularBuffer(3).is_empty()


def test_capacity():
    assert CircularBuffer(5).capacity() == 5


def test_push_and_full():
    b = CircularBuffer(3)
    b.push(1); b.push(2); b.push(3)
    assert b.is_full()
    assert b.peek() == 1


def test_fifo_pop():
    b = CircularBuffer(3)
    b.push(1); b.push(2); b.push(3)
    assert b.pop() == 1
    assert b.pop() == 2
    assert b.pop() == 3
    assert b.is_empty()


def test_wrap_around():
    b = CircularBuffer(3)
    b.push(1); b.push(2); b.push(3)
    b.pop(); b.pop()
    b.push(4); b.push(5)
    assert b.peek() == 3
    assert b.size() == 3


def test_push_full_raises():
    b = CircularBuffer(2)
    b.push(1); b.push(2)
    with pytest.raises(OverflowError):
        b.push(3)


def test_pop_empty_raises():
    with pytest.raises(IndexError):
        CircularBuffer(2).pop()


def test_zero_capacity_raises():
    with pytest.raises(ValueError):
        CircularBuffer(0)
